Other builders new format

rooooooooob opened the pull request:
Current status:

- [x] Other builders / utility code (rust)
- [x] Update other builder tests

To be done in later PRs
- tx_builder.rs (rust)
- WASM bindings
- Update all other tests

rooooooooob committed:
Move over builder code

rooooooooob committed:
Non-tx Builders refactored for new code

needed utility functionality like min ada calculation, conversions, etc
have been updated or added when needed.

rooooooooob commented:
Ignore the first commit when reviewing to see the diffs vs the old code. The first commit is just directly copying the old builder code.

SebastienGllmt commented on chain/rust/src/lib.rs:
```diff
@@ -347,6 +349,17 @@ pub struct Value {
     pub encodings: Option<ValueEncoding>,
 }
 
+// TODO: REMOVE after merging value/multiasset PR
```
TODO

gostkin replied:
we can merge already I beleive?

SebastienGllmt commented on chain/rust/src/crypto/utils.rs:
```diff
@@ -11,21 +18,23 @@
+        // let key = chain_crypto::PublicKey::<chain_crypto::ed25519_derive::Ed25519Bip32>::try_from(wit)?;
+        // let address_content = AddressContent::new_simple(Bip32PublicKey(key), protocol_magic);
+        // Ok(address_content)
+        todo!("redo byron address attributes to not be just a bytestring")
```
TODO

SebastienGllmt commented on chain/rust/src/crypto/hash.rs:
```diff
@@ -0,0 +1,136 @@
+                datum.serialize(&mut buf, true).unwrap();
+            }
+        }
+        todo!("language views encoding");
```
TODO

rooooooooob replied:
Done now.

SebastienGllmt commented on chain/rust/src/builders/witness_builder.rs:
```diff
@@ -0,0 +1,698 @@
+pub fn merge_fake_witness(builder: &mut TransactionWitnessSetBuilder, required_wits: &RequiredWitnessSet) {
+    let mut remaining_wits = required_wits.clone();
+    builder.vkeys.keys().for_each(|key| { remaining_wits.vkeys.remove(&key.hash()); });
+    todo!("need to adjust byron addr attributes to not just be bytes");
```
TODO

SebastienGllmt commented on chain/rust/src/builders/withdrawal_builder.rs:
```diff
@@ -0,0 +1,119 @@
+
+    pub fn plutus_script(&self, partial_witness: &PartialPlutusWitness, required_signers: &RequiredSigners) -> Result<WithdrawalBuilderResult, WithdrawalBuilderError> {
+        let mut required_wits = RequiredWitnessSet::default();
+        todo!("the line below won't work until we regenerate RequiredSigners with https://github.com/dcSpark/cddl-codegen/issues/164 fixed");
```
TODO

SebastienGllmt commented on chain/rust/src/builders/withdrawal_builder.rs:
```diff
@@ -0,0 +1,119 @@
+pub enum WithdrawalBuilderError {
+    #[error("Missing the following witnesses for the withdrawal: {0:?}. ")]
+    MissingWitnesses(RequiredWitnessSet),
+    //#[error("Withdrawal required a script, not a payment key: {}", .to_address().to_bech32(None))]
```
duplicate line?

rooooooooob replied:
I forgot to put a comment in this github PR after asking. I wasn't able to figure out how to call functions in this auto-format string in the thiserror macro to mimic the previous error message (bech32). Is it okay to just debug display the address? Or would anyone happen to know how to print it like that? I guess I could always manually implement this error type and its traits instead.

SebastienGllmt replied:
not sure. Bech32 is not to have, but hex is better than nothing

SebastienGllmt commented on chain/rust/src/builders/utils.rs:
```diff
@@ -0,0 +1,126 @@
+/// since u64 in CBOR are fixed size (will prefix with 0s to stay a constant width regardless of value)
+pub const force_u64: u64 = 0x01_00_00_00_00; 
+
+// impl GeneralTransactionMetadata {
```
Why is all of this commented out?

rooooooooob replied:
Not referred to so far. I'll probably remove it after finishing migrating `tx_builder.rs` if it's not needed. That is unless we still want them. It just seemed like more of the helper functionality to help with WASM making many temporaries like the ones we talked about in the other builders.

SebastienGllmt commented on chain/rust/src/builders/mod.rs:
```diff
@@ -0,0 +1,9 @@
+pub mod input_builder;
+pub mod redeemer_builder;
+pub mod output_builder;
+//pub mod tx_builder;
```
is this meant to be commented out?

rooooooooob replied:
For now, yes. The rest of the builders all compile since they've been updated. The tx_builder.rs does not yet. I will uncomment this once that's ready. It's a draft PR still.

SebastienGllmt commented on chain/rust/src/builders/mint_builder.rs:
```diff
@@ -0,0 +1,61 @@
+        let mut required_wits = RequiredWitnessSet::default();
+
+        let script_hash = partial_witness.script.hash();
+        todo!("the line below won't work until we regenerate RequiredSigners with https://github.com/dcSpark/cddl-codegen/issues/164 fixed");
```
TODO

SebastienGllmt commented on chain/rust/src/builders/input_builder.rs:
```diff
@@ -0,0 +1,135 @@
+
+    pub fn plutus_script(&self, partial_witness: &PartialPlutusWitness, required_signers: &RequiredSigners, datum: &PlutusData) -> Result<InputBuilderResult, InputBuilderError> {
+        let mut required_wits = RequiredWitnessSet::default();
+        todo!("the line below won't work until we regenerate RequiredSigners with https://github.com/dcSpark/cddl-codegen/issues/164 fixed");
```
TODO

SebastienGllmt commented on chain/rust/src/builders/certificate_builder.rs:
```diff
@@ -0,0 +1,202 @@
+
+    pub fn plutus_script(&self, partial_witness: &PartialPlutusWitness, required_signers: &RequiredSigners) -> Result<CertificateBuilderResult, CertBuilderError> {
+        let mut required_wits = RequiredWitnessSet::default();
+        todo!("the line below won't work until we regenerate RequiredSigners with https://github.com/dcSpark/cddl-codegen/issues/164 fixed");
```
TODO

gostkin commented on chain/rust/src/builders/mint_builder.rs:
```diff
@@ -0,0 +1,61 @@
+
+#[derive(Clone)]
+pub struct MintBuilderResult {
+    pub(crate) policy_id: PolicyId,
```
why `pub(crate)`? I think it makes sense to allow access to the fields

gostkin commented on chain/rust/src/builders/output_builder.rs:
```diff
@@ -0,0 +1,147 @@
+
+#[derive(Clone, Debug)]
+pub struct SingleOutputBuilderResult {
+    output: TransactionOutput,
```
pub?

gostkin commented on chain/rust/src/builders/input_builder.rs:
```diff
@@ -0,0 +1,135 @@
+
+#[derive(Clone, Debug)]
+pub struct InputBuilderResult {
+    pub(crate) input: TransactionInput,
```
Why pub(crate)? there's no way to access fields externally when you get this result, it's not very convenient

gostkin commented on chain/rust/src/builders/certificate_builder.rs:
```diff
@@ -0,0 +1,202 @@
+
+#[derive(Clone)]
+pub struct CertificateBuilderResult {
+    pub(crate) cert: Certificate,
```
same question about `pub(crate)`

gostkin commented on chain/rust/src/min_ada.rs:
```diff
@@ -0,0 +1,397 @@
+    ArithmeticOverflow,
+}
+
+// /// Provide backwards compatibility to Alonzo by taking the max min value of both er
```
is this part needed?

rooooooooob replied:
@SebastienGllmt do we need any previous era functionality that isn't just parsing on-chain block cbor?

gostkin commented on chain/rust/src/builders/witness_builder.rs:
```diff
@@ -0,0 +1,698 @@
+pub struct RequiredWitnessSet {
+    // note: the real key type for these is Vkey
+    // but cryptographically these should be equivalent and Ed25519KeyHash is more flexible
+    pub(crate) vkeys: BTreeSet<Ed25519KeyHash>,
```
What do you think about allowing the access to the fields?

rooooooooob replied:
Done as discussed.

gostkin commented on chain/rust/src/builders/withdrawal_builder.rs:
```diff
@@ -0,0 +1,119 @@
+
+#[derive(Clone)]
+pub struct WithdrawalBuilderResult {
+    pub(crate) address: RewardAddress,
```
same question

gostkin commented on chain/rust/src/builders/tx_builder.rs:
```diff
@@ -0,0 +1,4966 @@
+}
+
+#[wasm_bindgen]
+pub enum CoinSelectionStrategyCIP2 {
```
what do you think if we divide the file to multiple ones: e.g. separate file per struct. it's too big now

rooooooooob replied:
The vast majority of the file is for tests. ~1500 is the actual code. It might shrink once rewritten for idiomatic rust code. Here's the approximate breakdown:
* add_change_if_needed: ~300 (free-floating pub function that mutates the tx builder)
* SignedTxBuilder: ~40
* TxRedeemerBuilder: ~80

I don't think per-struct makes sense for breaking it up since the non-tx-builder structs are tiny. You'd have to start breaking off methods off of the tx builder struct itself to make it meaningfully smaller. The coin selection functions are ~250 lines for example.

gostkin commented on chain/rust/src/builders/tx_builder.rs:
```diff
@@ -0,0 +1,4966 @@
+    key_deposit: Option<BigNum>,       // protocol parameter
+    max_value_size: Option<u32>,       // protocol parameter
+    max_tx_size: Option<u32>,          // protocol parameter
+    /// TODO: remove once Babbage is on mainnet
```
babbage is on mainnet already

SebastienGllmt replied:
yeah, this come from the old code that was copy-pasted. No need to change it in this PR

gostkin replied:
@SebastienGllmt , we can do it later, imo makes sense to get rid of outdated functionality

gostkin replied:
issue for that -- https://github.com/dcSpark/cardano-multiplatform-lib/issues/200

gostkin commented on chain/rust/src/builders/tx_builder.rs:
```diff
@@ -0,0 +1,4966 @@
+
+#[wasm_bindgen]
+#[derive(Clone, Debug, Default)]
+pub struct TransactionBuilderConfigBuilder {
```
could we add impls for mainnet / testnets with a note that the params may change? and the descriptions where to get actual parameters. now it's not super friendly for end users (which might have no idea about cardano parameters)

SebastienGllmt replied:
The parameters can change over time which is why we hesitated to add defaults to CML

gostkin replied:
they can change, but it's not frequent as far as I see. why not to simplify the setup? (of course saying that the parameters can change and keeping the way to configure them manually)

rooooooooob replied:
How many people using CML would actually pull from up to date info instead? I wonder how many people would, instead, copy and paste parameters they got somewhere and were up to date when they made it but then go out of date later. I guess it shifts responsibility from us to the users.

gostkin replied:
> How many people using CML would actually pull from up to date info instead? I wonder how many people would, instead, copy and paste parameters they got somewhere and were up to date when they made it but then go out of date later. I guess it shifts responsibility from us to the users.

if we explain somewhere in the docs how to get actual parameters / provide tool for that, imo would be great as well from ux perspective

gostkin replied:
e.g. at least give a link to https://cexplorer.io/params in the docs

rooooooooob replied:
Looking at that graph there's some periods of time where the parameters were changing more often than I had expected. I think I'm more on the side of not providing it now, even if there are a lot of periods where it's pretty stable. Improving the docs about it is a great idea though.

gostkin commented on specs/babbage/transaction.cddl:
```diff
@@ -41,7 +41,7 @@ alonzo_tx_out = [
 ]
 babbage_tx_out = {
     0 : address,      ; @name address
-    1 : value,        ; @name value
```
why value -> amount?

rooooooooob replied:
For consistency, to match the name used in the other 2 variants.

rooooooooob committed:
Finish non-tx_builder builders main work

* regen BootstrapWitness with updated cddl that refers to byron code
  instead of opaque bytes + finish witness builder that needed it

* regen required_signers after cddl-codegen fix + builder todos related

* A few APIs were still cloning and taking references. changed to take
  ownership

* change pub(crate) to pub in several places for rust only

rooooooooob commented on chain/rust/src/builders/redeemer_builder.rs:
```diff
@@ -0,0 +1,330 @@
+        
+    }
+
+    pub fn add_cert(&mut self, result: &CertificateBuilderResult) {
```
I noticed this (as well as `TransactionWitnessSetBuilder`) are different from the other builders API-wise. @SebastienGllmt do you remember if there was a reason for this? They modify and don't return, whereas other ones return self to allow for function chaining. Was there a performance difference on WASM due to all the cloning instead of moving like in rust? Or some other reason?

SebastienGllmt replied:
I don't recall any deep reason. These were originally meant to be Rust-only I think but I haven't double-checked how they will be used going forward

rooooooooob committed:
Script data hash / Language views + related fixes

Required fixes that caused hash to not work:

1) Regen serialization code

Includes fixes made with cddl-codegen#176

This prevents deserialization confusion between shelley and alonzo tx
outs as their first 2 fields are identical.

and also with cddl-codegen#179 to get proper read_len checks when
optional fields are external types (incl raw bytes encoding)

2) Constructor plutus data encoding didn't remember which encoding.

It seems that compact-sized constructors can also be encoded in the general format.

rooooooooob committed:
Merge branch 'develop' into builders-new-format

rooooooooob committed:
Update non-tx-builder builder tests

Includes moving over + updating icarus/daedalus witness builder utils to
work with the new code/new tests.

gostkin commented:
btw @rooooooooob can we make `cargo build --release` compiling for the root project? I see we're missing `chain/genesis` for that

gostkin commented:
> btw @rooooooooob can we make `cargo build --release` compiling for the root project? I see we're missing `chain/genesis` for that

If we resolve this we can enable a compile check at least

rooooooooob commented:
>btw @rooooooooob can we make cargo build --release compiling for the root project? I see we're missing chain/genesis for that

Is that in scope here? We didn't touch any of those files. I could do another PR that updates non-builder tests to compile + gets that genesis to compile on release so we can get the compile + tests (if it's not too much work) checks working there.

gostkin commented:
> > btw @rooooooooob can we make cargo build --release compiling for the root project? I see we're missing chain/genesis for that
> 
> Is that in scope here? We didn't touch any of those files. I could do another PR that updates non-builder tests to compile + gets that genesis to compile on release so we can get the compile + tests (if it's not too much work) checks working there.

no, that's not in the scope of builders pr
